finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4ServerDispatcher.scala

```diff
@@ -85,11 +85,17 @@ class Netty4ServerDispatcher(
     val serveF = st.onRecvMessage.flatMap(serve).flatMap(st.send(_).flatten)
 
     // When the stream is reset, ensure that the cancelation is
-    // propagated downstream.
+    // propagated downstream and discard the recv message (if we have one).
     st.onReset.onFailure {
-      case StreamError.Remote(rst: Reset) => serveF.raise(rst)
-      case StreamError.Remote(e) => serveF.raise(Reset.Cancel)
-      case e => serveF.raise(e)
+      case StreamError.Remote(rst: Reset) =>
+        st.onRecvMessage.onSuccess(_.stream.cancel(rst))
```

Comment:
Is there a situation where there may be more frames coming in for the message, but they are not there yet? If so, is `_.stream.cancel(rst))` still able to handle that?

Reply:
Good question.  `stream.cancel` first fails the underlying queue so that it cannot accept any new frame, before draining that queue.